common-pipeline-steps/src/main/scala/com/acxiom/pipeline/steps/MappingSteps.scala

```diff
@@ -0,0 +1,239 @@
+    "Pipeline"
+  )
+  def mergeDataFrames(inputDataFrame: DataFrame, destinationDataFrame: DataFrame, mappings: Mappings = Mappings(List()),
+                      addNewColumns: Boolean = true): DataFrame = {
+    // map to destination dataframe
+    val mappedFromDF = mapToDestinationDataFrame(inputDataFrame, destinationDataFrame, mappings, addNewColumns)
+    // treating 2 as the driver...adding attributes from df1 that don't exist on df2
+    val finalToDF = addMissingDestinationAttributes(destinationDataFrame, mappedFromDF.schema)
+    // union dataframes together
+    finalToDF.union(mappedFromDF)
+  }
+
+  /**
+    * applies transform logic to override existing fields or append new fields to the end of an existing dataframe
+    * @param dataFrame  the input dataframe
+    * @param mappings list of transform object containing expressions to generate fields
+    * @return   an updated dataframe
+    */
+  @StepFunction(
+    "ac3dafe4-e6ee-45c9-8fc6-fa7f918cf4f2",
+    "Modify or Create Columns using Transforms Provided",
+    "This step transforms existing columns and/or adds new columns to an existing dataframe using expressions provided",
+    "Pipeline"
+  )
+  def applyTransforms(dataFrame: DataFrame, mappings: Mappings): DataFrame = {
+    // pull out mappings that contain a transform
+    val mappingsWithTransforms = mappings.details.filter(_.transform.nonEmpty)
+
+    // apply any alias logic to input column names
+    val aliasedDF = applyAliasesToInputDataFrame(dataFrame, mappings)
+
+    // create input dataframe with any transform overrides (preserving input order)
+    val inputExprs = aliasedDF.columns.map(a => {
+      val mapping = mappingsWithTransforms.find(_.outputField == a)
+      if(mapping.nonEmpty) {
+        logger.info(s"adding transform for existing column '${mapping.get.outputField}', transform=${mapping.get.transform.get}")
+        expr(mapping.get.transform.get).as(mapping.get.outputField)
+      } else { col(a) }
+    })
+
+    // append transforms creating new fields to the end
+    val finalExprs = mappingsWithTransforms.foldLeft(inputExprs)( (exprList, m) => {
+      if(aliasedDF.columns.contains(m.outputField)) {
+        exprList
+      } else {
+        logger.info(s"adding transform for new column '${m.outputField},transform=${m.transform.get}")
+        exprList :+ expr(m.transform.get).as(m.outputField)
+      }
+    })
+
+    // return dataframe with all transforms applied
+    aliasedDF.select(finalExprs: _*)
+  }
+
+  /**
+    * converts data types from one dataframe to match data types of destination schema
+    * @param dataFrame          the input dataframe
+    * @param destinationSchema  the schema with the data types to adhere to
+    * @return   a new dataframe where all datatypes match destination schema data types
+    */
+  def convertDataTypesToDestination(dataFrame: DataFrame, destinationSchema: StructType): DataFrame = {
```

Comment:
Would it be useful to expose the remaining functions in this step object

Reply:
I thought about that, but in the end figured that if the user needed any of this functionality, they would just use one of the 'mapDataFrame' methods that are exposed.  For example, if you want to convertDataTypes, no need to run that method directly, just run mapDataFrameToSchema() or maptoDataFrame() and that step will be covered.

Reply:
I made those private to the steps package in the latest commit